File: Java/src/test/java/com/cryptofundamentals/AsymmetricEncryptionTest.java
package com.cryptofundamentals;

import static org.junit.Assert.assertArrayEquals;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

import java.security.KeyPair;
import java.security.PrivateKey;
import java.security.PublicKey;

import javax.crypto.SecretKey;

import org.junit.Test;

public class AsymmetricEncryptionTest {

    @Test
    public void testGenerateRSAKeyPair() throws Exception {
        KeyPair keyPair = Crypto.generateRsaKey();

        assertEquals("RSA", keyPair.getPublic().getAlgorithm());
        assertTrue(keyPair.getPublic().getEncoded().length > 2048 / 8);
        assertTrue(keyPair.getPrivate().getEncoded().length > 2048 / 8);
    }

    @Test
    public void testEncryptSymmetricKey() throws Exception {
        KeyPair keyPair = Crypto.generateRsaKey();

        PublicKey publicKey = keyPair.getPublic();
        PrivateKey privateKey = keyPair.getPrivate();

        SecretKey key = Crypto.generateAesKey();

        byte[] encryptedKey = Crypto.encryptWithRsa(publicKey, key);
        byte[] decryptedKey = Crypto.decryptWithRsa(privateKey, encryptedKey);

        assertArrayEquals(key.getEncoded(), decryptedKey);
    }

    @Test
    public void testSignMessage() throws Exception {
        KeyPair keyPair = Crypto.generateRsaKey();

        PublicKey publicKey = keyPair.getPublic();
        PrivateKey privateKey = keyPair.getPrivate();

        String message = "Alice knows Bob's secret.";
        byte[] messageBytes = message.getBytes();

        byte[] signatureBytes = Crypto.signMessage(privateKey, messageBytes);
        boolean verified = Crypto.verifySignature(publicKey, messageBytes, signatureBytes);

        assertTrue(verified);
    }
}

File: .NET/CryptoHelpers/CryptoHelpers.Test/AsymmetricEncryptionTest.cs
using FluentAssertions;
using System.Security.Cryptography;
using System.Text;
using Xunit;

namespace CryptoHelpers.Test
{
    public class AsymmetricEncryptionTest
    {
        [Fact]
        public void GenerateRsaKeyPair()
        {
            var rsa = RSA.Create(2048);
            int keySize = rsa.KeySize;
            keySize.Should().Be(2048);
        }

        [Fact]
        public void SharePublicKey()
        {
            var rsa = RSA.Create(2048);
            byte[] publicKey = rsa.ExportRSAPublicKey();
            publicKey.Length.Should().Be((2048 + 24 + 88) / 8);
        }

        [Fact]
        public void EncryptSymmetricKey()
        {
            var rsa = RSA.Create(2048);
            var publicKey = AsymmetricHelpers.WritePEM("RSA PUBLIC KEY", rsa.ExportRSAPublicKey());

            var key = SymmetricAlgorithm.Create("AES").Key;

            byte[] encryptedKey = AsymmetricHelpers.EncryptWithRsa(publicKey, key);
            byte[] decryptedKey = AsymmetricHelpers.DecryptWithRsa(rsa, encryptedKey);

            decryptedKey.Should().Equal(key);
        }

        [Fact]
        public void SignMessage()
        {
            var rsa = RSA.Create(2048);
            var publicKey = AsymmetricHelpers.WritePEM("RSA PUBLIC KEY", rsa.ExportRSAPublicKey());

            string message = "Alice knows Bob's secret.";
            byte[] messageBytes = Encoding.UTF8.GetBytes(message);

            byte[] signatureBytes = AsymmetricHelpers.SignMessage(rsa, messageBytes);
            bool verified = AsymmetricHelpers.VerifySignature(publicKey, messageBytes, signatureBytes);

            verified.Should().BeTrue();
        }
    }
}
